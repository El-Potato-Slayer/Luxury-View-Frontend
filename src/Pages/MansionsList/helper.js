import { Fragment } from 'react';
import { Link } from 'react-router-dom';
import { mainRooms, priceOutput } from '../../Helpers/helper';

function listRooms(mansion) {
  return (
    <p className="listing-room">
      {mainRooms(mansion).map((room) => (
        <Fragment key={room.id}>
          <span>
            {room.amount}
            {' '}
            {room.name}
            {' '}
            &bull;
          </span>
        &nbsp;
        </Fragment>
      ))}
      <span>
        {mansion.land_area}
        m
        <sup>2</sup>
      </span>
    </p>
  );
}

export function listMansionDetails(mansion) {
  return (
    <Link to={`/mansions/${mansion.id}`}>
      <div>
        <div
          className="listing-image"
          style={{ background: `url(${mansion.picture}) center / cover` }}
        />
        <div className="info">
          <p className="listing-important">
            $&nbsp;
            {priceOutput(mansion.price)}
          </p>
          <p>{mansion.name}</p>
          {listRooms(mansion)}
        </div>
      </div>
    </Link>
  );
}

export function listAgent(mansion) {
  return (
    <Link to={`agents/${mansion.agent.id}`} className="listing-agent">
      <p>
        <span>{mansion.agent.first_name}</span>
        &nbsp;
        <span>{mansion.agent.last_name}</span>
      </p>
      <img src={mansion.agent.picture} alt="agent" />
    </Link>
  );
}
