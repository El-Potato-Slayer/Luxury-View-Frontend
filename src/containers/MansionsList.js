import PropTypes from 'prop-types';
import { useEffect } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

function MansionsList({ mansions }) {
  // const [rooms, setRooms] = useState([]);
  useEffect(() => {
    // console.log(mansions[0].rooms);
    // setRooms(mansions.rooms);
    // console.log(rooms);
  }, []);

  const filteredRooms = (mansion) => mansion.rooms.filter((room) => (
    room.name === 'Bedrooms' || room.name === 'Bathrooms'
  ));

  function listRooms(mansion) {
    return (
      <p>
        {filteredRooms(mansion).map((room) => (
          <>
            <span>
              {room.amount}
              {' '}
              {room.name}
              {' '}
              &bull;
            </span>
          &nbsp;
          </>
        ))}
        <span>
          {mansion.land_area}
          m
          <sup>2</sup>
        </span>
      </p>
    );
  }

  function listMansionDetails(mansion) {
    return (
      <Link to={`/mansions/${mansion.id}`}>
        <img src={mansion.picture} alt="" />
        <p>
          $
          {mansion.price}
        </p>
        <p>{mansion.name}</p>
        {listRooms(mansion)}
      </Link>
    );
  }

  function listAgent(mansion) {
    return (
      <Link to={`agents/${mansion.agent.id}`} className="mansion-list-agent">
        <span>{mansion.agent.first_name}</span>
        &nbsp;
        <span>{mansion.agent.last_name}</span>
        <img src={mansion.agent.picture} alt="agent" />
      </Link>
    );
  }

  return (
    <>
      <h2 data-testid="mansions">Mansions</h2>
      <div data-testid="mansions-listing">
        {mansions.map((mansion) => (
          <div key={mansion.id}>
            {listMansionDetails(mansion)}
            {listAgent(mansion)}
          </div>
        ))}
      </div>
    </>
  );
}

const mapStateToProps = (state) => ({
  mansions: state.propertiesReducer.properties,
});

MansionsList.propTypes = {
  mansions: PropTypes.instanceOf(Array).isRequired,
};

export default connect(mapStateToProps)(MansionsList);
