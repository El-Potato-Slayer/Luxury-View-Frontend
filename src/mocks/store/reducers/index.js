import { combineReducers } from 'redux';
import agentsReducer from './agentsReducer';
import appointmentsReducer from './appointmentsReducer';
import authReducer from './authReducer';
import propertiesReducer from './propertiesReducer';
import userReducer from './userReducer';
import filterReducer from './filterReducer';

export default combineReducers({
  agentsReducer, appointmentsReducer, authReducer, propertiesReducer, userReducer, filterReducer,
});
