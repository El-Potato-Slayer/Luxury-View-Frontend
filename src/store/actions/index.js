export const setAgents = (payload) => ({
  type: 'SET_AGENTS',
  payload,
});

export const setAppointments = (payload) => ({
  type: 'SET_APPOINTMENTS',
  payload,
});

export const setProperties = (payload) => ({
  type: 'SET_PROPERTIES',
  payload,
});

export const changeFilter = (name) => ({
  type: 'CHANGE_FILTER',
  name,
});
