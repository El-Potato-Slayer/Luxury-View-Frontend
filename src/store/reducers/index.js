import { combineReducers } from 'redux';
import agentsReducer from './agentsReducer';
import appointmentsReducer from './appointmentsReducer';
import authReducer from './authReducer';
import propertiesReducer from './propertiesReducer';
import userReducer from './userReducer';

export default combineReducers({
  agentsReducer, appointmentsReducer, authReducer, propertiesReducer, userReducer,
});
