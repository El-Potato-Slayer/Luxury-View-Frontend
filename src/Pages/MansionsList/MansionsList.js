import { useSelector } from 'react-redux';
import SkeletonListItem from '../../Skeletons/SkeletonListMansion/SkeletonListMansion';
import { sortMansionByNewest } from '../../Helpers/index';
import { listAgent, listMansionDetails } from './helper';

function MansionsList() {
  const { properties: mansions, loading } = useSelector((state) => state.propertiesReducer);

  return (
    <div data-testid="mansions" className="page">
      <h2 className="page-title">Mansions</h2>
      <div data-testid="listings" className="listings">
        {loading && (
          <>
            <SkeletonListItem />
            <SkeletonListItem />
            <SkeletonListItem />
          </>
        )}
        {sortMansionByNewest(mansions).map((mansion) => (
          <div key={mansion.id}>
            {listMansionDetails(mansion)}
            {listAgent(mansion)}
          </div>
        ))}
      </div>
    </div>
  );
}

export default MansionsList;
