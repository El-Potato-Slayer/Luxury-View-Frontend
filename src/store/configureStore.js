import { applyMiddleware, createStore } from 'redux';
import ReduxThunk from 'redux-thunk';
import { fetchData } from './actions';
import { fetchAgentsFailure, fetchAgentsRequest, fetchAgentsSuccess } from './actions/agentActions';
import { fetchAppointmentsFailure, fetchAppointmentsRequest, fetchAppointmentsSuccess } from './actions/appointmentActions';
import { fetchPropertiesFailure, fetchPropertiesRequest, fetchPropertiesSuccess } from './actions/propertyActions';
import rootReducer from './reducers';

const store = createStore(rootReducer, applyMiddleware(ReduxThunk));

// store.subscribe(() => { console.log(store.getState()); });
store.dispatch(fetchData('properties', fetchPropertiesRequest, fetchPropertiesSuccess, fetchPropertiesFailure));
store.dispatch(fetchData('agents', fetchAgentsRequest, fetchAgentsSuccess, fetchAgentsFailure));
store.dispatch(fetchData('appointments', fetchAppointmentsRequest, fetchAppointmentsSuccess, fetchAppointmentsFailure, true));

export default store;
