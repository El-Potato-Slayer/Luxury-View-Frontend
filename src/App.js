import './App.css';
import {
  BrowserRouter, Redirect, Route, Switch,
} from 'react-router-dom';
import { connect, useDispatch } from 'react-redux';
import { useEffect } from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import Navbar from './components/Navbar';
import Home from './components/Home';
import { setAgents, setAuth, setProperties } from './store/actions';
import MansionsList from './containers/MansionsList';
import AgentsList from './containers/AgentsList';
import ProtectedRoute from './components/ProtectedRoute';
import AppointmentsList from './containers/AppointmentsList';
import Login from './components/Login';
import Mansion from './components/Mansion';

function App({ isAuth }) {
  const dispatch = useDispatch();
  let token;
  useEffect(() => {
    axios.get('http://localhost:3000/api/v1/properties')
      .then((resp) => {
        dispatch(setProperties(resp.data));
      });
    axios.get('http://localhost:3000/api/v1/agents')
      .then((resp) => {
        dispatch(setAgents(resp.data));
      });
    token = localStorage.getItem('token');
    if (token) {
      const tokenExpiration = JSON.parse(atob(token.split('.')[1])).exp;
      const dateNow = new Date();
      if (tokenExpiration > dateNow.getTime() / 1000) {
        dispatch(setAuth(true));
      }
    }
  }, []);
  return (
    <div data-testid="app" className="App">
      <BrowserRouter>
        <Navbar />
        <Switch>
          <Route exact path="/">
            <Home />
          </Route>
          <Route exact path="/mansions">
            <MansionsList />
          </Route>
          <Route exact path="/agents">
            <AgentsList />
          </Route>
          <Route path="/login">
            {isAuth ? <Redirect to="/appointments" /> : <Login />}
          </Route>
          <Route path="/mansions/:id">
            <Mansion />
          </Route>
          <ProtectedRoute exact path="/appointments" component={AppointmentsList} />
        </Switch>
      </BrowserRouter>
    </div>
  );
}

const mapStateToProps = (state) => ({
  isAuth: state.authReducer.isAuth,
});

App.propTypes = {
  isAuth: PropTypes.bool.isRequired,
};

export default connect(mapStateToProps)(App);
