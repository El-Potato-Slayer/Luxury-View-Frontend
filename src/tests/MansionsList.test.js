import {
  cleanup, fireEvent, render, screen,
} from '@testing-library/react';
import { BrowserRouter, Switch } from 'react-router-dom';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import MansionsList from '../containers/MansionsList';
import store from '../mocks/store/configureStore';
import App from '../App';

jest.mock('axios');
afterEach(cleanup);

test('Renders the MansionsList component without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <Provider store={store}>
      <BrowserRouter>
        <MansionsList />
      </BrowserRouter>
    </Provider>,
    div,
  );
});

test('Checks if title is rendered', () => {
  render(
    <Provider store={store}>
      <BrowserRouter>
        <MansionsList />
      </BrowserRouter>
    </Provider>,
  );
  expect(screen.getByTestId('mansions')).toHaveTextContent('Mansions');
});

test('Checks if mansions are rendered', () => {
  render(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <Switch>
            <MansionsList />
          </Switch>
        </BrowserRouter>
      </App>
    </Provider>,
  );
  fireEvent.click(screen.getByTestId('navbar-mansions'));
  expect(screen.getByTestId('mansions')).toHaveTextContent('Suganti Vapi');
  expect(screen.getByTestId('mansions')).toHaveTextContent('6 Bed Mansion');
  expect(screen.getByTestId('mansions')).toHaveTextContent('Modern Trophy Estate');
  expect(screen.getByTestId('mansions')).toHaveTextContent('Videa Casale');
});

test('Checks if rooms are rendered', () => {
  render(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <Switch>
            <MansionsList />
          </Switch>
        </BrowserRouter>
      </App>
    </Provider>,
  );
  fireEvent.click(screen.getByTestId('navbar-mansions'));
  expect(screen.getByTestId('mansions')).toHaveTextContent('6 Bedrooms');
  expect(screen.getByTestId('mansions')).toHaveTextContent('8 Bedrooms');
  expect(screen.getByTestId('mansions')).toHaveTextContent('7 Bedrooms');
  expect(screen.getByTestId('mansions')).toHaveTextContent('5 Bedrooms');
});

test('Checks if bathrooms are rendered', () => {
  render(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <Switch>
            <MansionsList />
          </Switch>
        </BrowserRouter>
      </App>
    </Provider>,
  );
  fireEvent.click(screen.getByTestId('navbar-mansions'));
  expect(screen.getByTestId('mansions')).toHaveTextContent('6 Bathrooms');
  expect(screen.getByTestId('mansions')).toHaveTextContent('11 Bathrooms');
  expect(screen.getByTestId('mansions')).toHaveTextContent('12 Bathrooms');
  expect(screen.getByTestId('mansions')).toHaveTextContent('8 Bathrooms');
});

test('Checks if agents are rendered', () => {
  render(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <Switch>
            <MansionsList />
          </Switch>
        </BrowserRouter>
      </App>
    </Provider>,
  );
  fireEvent.click(screen.getByTestId('navbar-mansions'));
  expect(screen.getByTestId('mansions')).toHaveTextContent('Joshua Door');
  expect(screen.getByTestId('mansions')).toHaveTextContent('Darlene');
  expect(screen.getByTestId('mansions')).toHaveTextContent('Yapi Kobustus');
});

test('Checks if prices are rendered', () => {
  render(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <Switch>
            <MansionsList />
          </Switch>
        </BrowserRouter>
      </App>
    </Provider>,
  );
  fireEvent.click(screen.getByTestId('navbar-mansions'));
  expect(screen.getByTestId('mansions')).toHaveTextContent('$ 100,000,000');
  expect(screen.getByTestId('mansions')).toHaveTextContent('$ 26,000,000');
  expect(screen.getByTestId('mansions')).toHaveTextContent('$ 35,000,000');
  expect(screen.getByTestId('mansions')).toHaveTextContent('$ 2,500,000');
});

test('Checks if prices are not rendered without commas', () => {
  render(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <Switch>
            <MansionsList />
          </Switch>
        </BrowserRouter>
      </App>
    </Provider>,
  );
  fireEvent.click(screen.getByTestId('navbar-mansions'));
  expect(screen.getByTestId('mansions')).not.toHaveTextContent('$ 100000000');
  expect(screen.getByTestId('mansions')).not.toHaveTextContent('$ 26000000');
  expect(screen.getByTestId('mansions')).not.toHaveTextContent('$ 35000000');
  expect(screen.getByTestId('mansions')).not.toHaveTextContent('$ 2500000');
});

it('matches snapshot', () => {
  const tree = renderer.create(
    <Provider store={store}>
      <App>
        <BrowserRouter>
          <MansionsList />
        </BrowserRouter>
      </App>
    </Provider>,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
