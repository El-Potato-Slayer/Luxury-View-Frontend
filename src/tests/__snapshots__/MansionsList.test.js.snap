// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="App"
  data-testid="app"
>
  <div
    className="navbar"
    data-testid="navbar"
  >
    <div
      aria-hidden="true"
      className="dark-filter hide"
      onClick={[Function]}
      onKeyDown={[Function]}
    />
    <div
      className="burger"
    >
      <button
        className="burger-wrapper"
        id="burger-wrapper"
        onClick={[Function]}
        type="button"
      >
        <div
          className="burger-menu"
          id="burger-menu"
        />
      </button>
    </div>
    <img
      alt=""
      className="nav-logo"
      src="/assets/logo.png"
    />
    <div
      className="nav-links"
    >
      <a
        className="nav-item"
        data-testid="navbar-homepage"
        href="/"
        onClick={[Function]}
      >
        Home
      </a>
      <a
        className="nav-item"
        data-testid="navbar-mansions"
        href="/mansions"
        onClick={[Function]}
      >
        Mansions
      </a>
      <a
        className="nav-item"
        data-testid="navbar-appointments"
        href="/appointments"
        onClick={[Function]}
      >
        Appointments
      </a>
      <a
        className="nav-item"
        data-testid="navbar-agents"
        href="/agents"
        onClick={[Function]}
      >
        Agents
      </a>
      <a
        className="nav-item"
        data-testid="navbar-login"
        href="/login"
        onClick={[Function]}
      >
        Login
      </a>
      <a
        className="nav-item"
        data-testid="navbar-register"
        href="/register"
        onClick={[Function]}
      >
        Register
      </a>
    </div>
    <div
      className="nav-social"
    />
  </div>
  <div
    className="page"
    data-testid="mansions"
  >
    <h2
      className="page-title"
    >
      Mansions
    </h2>
    <div
      className="listings"
      data-testid="listings"
    >
      <div>
        <a
          href="/mansions/4"
          onClick={[Function]}
        >
          <div>
            <div
              className="listing-image"
              style={
                Object {
                  "background": "url(Mansion 4 Picture) center / cover",
                }
              }
            />
            <div
              className="info"
            >
              <p
                className="listing-important"
              >
                $ 
                2,500,000
              </p>
              <p>
                Modern Trophy Estate
              </p>
              <p
                className="listing-room"
              >
                <span>
                  5
                   
                  Bedrooms
                   
                  •
                </span>
                 
                <span>
                  8
                   
                  Bathrooms
                   
                  •
                </span>
                 
                <span>
                  735
                  m
                  <sup>
                    2
                  </sup>
                </span>
              </p>
            </div>
          </div>
        </a>
        <a
          className="listing-agent"
          href="/agents/3"
          onClick={[Function]}
        >
          <p>
            <span>
              Yapi
            </span>
             
            <span>
              Kobustus
            </span>
          </p>
          <img
            alt="agent"
            src="Picture of agent"
          />
        </a>
      </div>
      <div>
        <a
          href="/mansions/3"
          onClick={[Function]}
        >
          <div>
            <div
              className="listing-image"
              style={
                Object {
                  "background": "url(Mansion 3 Picture) center / cover",
                }
              }
            />
            <div
              className="info"
            >
              <p
                className="listing-important"
              >
                $ 
                35,000,000
              </p>
              <p>
                Suganti Vapi
              </p>
              <p
                className="listing-room"
              >
                <span>
                  7
                   
                  Bedrooms
                   
                  •
                </span>
                 
                <span>
                  12
                   
                  Bathrooms
                   
                  •
                </span>
                 
                <span>
                  1325
                  m
                  <sup>
                    2
                  </sup>
                </span>
              </p>
            </div>
          </div>
        </a>
        <a
          className="listing-agent"
          href="/agents/2"
          onClick={[Function]}
        >
          <p>
            <span>
              Darlene (unisex name)
            </span>
             
            <span>
              Memphis
            </span>
          </p>
          <img
            alt="agent"
            src="picture link"
          />
        </a>
      </div>
      <div>
        <a
          href="/mansions/2"
          onClick={[Function]}
        >
          <div>
            <div
              className="listing-image"
              style={
                Object {
                  "background": "url(Mansion 2 Picture) center / cover",
                }
              }
            />
            <div
              className="info"
            >
              <p
                className="listing-important"
              >
                $ 
                26,000,000
              </p>
              <p>
                Videa Casale
              </p>
              <p
                className="listing-room"
              >
                <span>
                  8
                   
                  Bedrooms
                   
                  •
                </span>
                 
                <span>
                  11
                   
                  Bathrooms
                   
                  •
                </span>
                 
                <span>
                  1031
                  m
                  <sup>
                    2
                  </sup>
                </span>
              </p>
            </div>
          </div>
        </a>
        <a
          className="listing-agent"
          href="/agents/2"
          onClick={[Function]}
        >
          <p>
            <span>
              Darlene (unisex name)
            </span>
             
            <span>
              Memphis
            </span>
          </p>
          <img
            alt="agent"
            src="Agent Picture Link"
          />
        </a>
      </div>
      <div>
        <a
          href="/mansions/1"
          onClick={[Function]}
        >
          <div>
            <div
              className="listing-image"
              style={
                Object {
                  "background": "url(Mansion 1 Picture) center / cover",
                }
              }
            />
            <div
              className="info"
            >
              <p
                className="listing-important"
              >
                $ 
                100,000,000
              </p>
              <p>
                6 Bed Mansion
              </p>
              <p
                className="listing-room"
              >
                <span>
                  6
                   
                  Bedrooms
                   
                  •
                </span>
                 
                <span>
                  6
                   
                  Bathrooms
                   
                  •
                </span>
                 
                <span>
                  1511
                  m
                  <sup>
                    2
                  </sup>
                </span>
              </p>
            </div>
          </div>
        </a>
        <a
          className="listing-agent"
          href="/agents/1"
          onClick={[Function]}
        >
          <p>
            <span>
              Joshua
            </span>
             
            <span>
              Door
            </span>
          </p>
          <img
            alt="agent"
            src="Agent Picture Link"
          />
        </a>
      </div>
    </div>
  </div>
</div>
`;
